include/tvm/relay/transform.h

```diff
@@ -83,18 +117,81 @@ class PassContextNode : public RelayNode {
    */
   ErrorReporter err_reporter;
 
+  /*! \brief The default optimization level. */
+  int opt_level{2};
+
+  /*! \brief CPU is the default fallback device for heterogeneous execution. */
+  int fallback_device{static_cast<int>(kDLCPU)};
+
+  /*! \brief The list of required passes. */
+  tvm::Array<tvm::Expr> required_pass;
+  /*! \brief The list of disabled passes. */
+  tvm::Array<tvm::Expr> disabled_pass;
+
+  /*! 
+   * \brief A helper struct to get the optimization pass name to opt level
+   * mapping.
+   */
+  OptPassLevel OPT_PASS_LEVEL;
+
+  /*!
+   * \brief Convert a list of tvm StringImm to a `std::string` set.
+   *
+   * \param input. The input StringImm array.
+   *
+   * \return The coverted `std::strin`g set.
+   */
+  std::unordered_set<std::string> ToStringSet(
+      const tvm::Array<tvm::Expr>& input) const;
+
+  /*!
+   * \brief Check if a pass is enabled.
+   *
+   * \param pass_name The name of an optimization/analysis pass.
+   *
+   * \return true if the pass is enabled. Otherwise, false.
+   */
+  bool pass_enabled(const std::string& pass_name) const;
+
   PassContextNode() = default;
 
   void VisitAttrs(tvm::AttrVisitor* v) final {
+    v->Visit("opt_level", &opt_level);
+    v->Visit("fallback_device", &fallback_device);
+    v->Visit("required_pass", &required_pass);
+    v->Visit("disabled_pass", &disabled_pass);
   }
 
-  TVM_DLL static PassContext make();
-
   static constexpr const char* _type_key = "relay.PassContext";
   TVM_DECLARE_NODE_TYPE_INFO(PassContextNode, RelayNode);
 };
 
-TVM_DEFINE_NODE_REF(PassContext, PassContextNode)
+class PassContext : public NodeRef {
+ public:
+  PassContext() {}
+  explicit PassContext(tvm::NodePtr<Node> n) : NodeRef(n) {}
+
+  TVM_DLL PassContext(int opt_level, int fallback_device,
+                      tvm::Array<tvm::Expr> required_pass,
+                      tvm::Array<tvm::Expr> disabled_pass);
+
+  // The entry of a pass context scope.
+  TVM_DLL static void EnterWithScope(const PassContext& pass_ctx);
```

Comment:
ENTER/exit declare as private, and use friend class

Reply:
Will do when #3231 is merged.